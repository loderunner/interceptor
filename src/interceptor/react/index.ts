export { Provider as InterceptProvider, useIntercepts } from './intercept'
