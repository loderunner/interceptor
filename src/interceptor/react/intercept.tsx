import * as React from 'react'
import { createContext, useContext, useState } from 'react'

import * as intercept from '../intercept'

interface InterceptContextType {
  intercepts: intercept.Intercept[]
  addIntercept: (pattern: string) => void
  removeIntercept: (inter: intercept.Intercept) => void
}

const InterceptContext = createContext<InterceptContextType>({
  intercepts: [],
  addIntercept: () => {},
  removeIntercept: () => {},
})

interface Props {
  children: React.ReactNode
}

export const Provider = ({ children }: Props) => {
  const [intercepts, setIntercepts] = useState([...intercept.intercepts])

  const addIntercept = (pattern: string) => {
    intercept.addIntercept({ pattern, enabled: true })
    setIntercepts([...intercept.intercepts])
  }
  const removeIntercept = (inter: intercept.Intercept) => {
    intercept.removeIntercept(inter)
    setIntercepts([...intercept.intercepts])
  }

  return (
    <InterceptContext.Provider
      value={{ intercepts, addIntercept, removeIntercept }}
    >
      {children}
    </InterceptContext.Provider>
  )
}

export const useIntercepts = () => useContext(InterceptContext)
