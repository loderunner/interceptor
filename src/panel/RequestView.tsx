import * as cookie from 'cookie'
import * as React from 'react'
import { useCallback, useLayoutEffect, useMemo, useRef, useState } from 'react'

import {
  Copy as CopyIcon,
  UnfoldLess as UnfoldLessIcon,
  UnfoldMore as UnfoldMoreIcon,
} from '@/icons'
import { useRequest } from '@/interceptor/hooks'

const SectionValue = ({ value }: { value: string }) => {
  const [foldable, setFoldable] = useState(false)
  const [folded, setFolded] = useState(true)
  const [showButtons, setShowButtons] = useState(false)
  const spanRef = useRef<HTMLSpanElement>(null)

  useLayoutEffect(() => {
    if (spanRef.current) {
      setFoldable(spanRef.current.scrollWidth > spanRef.current.clientWidth)
    }
  }, [spanRef])

  const className = useMemo(() => {
    if (folded) {
      return 'text-ellipsis overflow-x-hidden whitespace-nowrap'
    } else {
      return 'break-all'
    }
  }, [folded])

  const onFold = useCallback(() => setFolded(true), [])
  const onUnfold = useCallback(() => setFolded(false), [])

  const buttonClassName = useMemo(() => {
    let className =
      'h-4 w-4 p-0.5 rounded-sm bg-slate-400 hover:bg-slate-500 active:bg-slate-600'
    if (!showButtons) {
      className += ' invisible'
    }
    return className
  }, [showButtons])

  const unfoldButton = useMemo(() => {
    if (!foldable) {
      return null
    }

    return (
      <button className={buttonClassName}>
        {folded ? (
          <UnfoldMoreIcon className="fill-white" onClick={onUnfold} />
        ) : (
          <UnfoldLessIcon className="fill-white" onClick={onFold} />
        )}
      </button>
    )
  }, [foldable, folded, buttonClassName, onFold, onUnfold])

  const onHoverStart = useCallback(() => setShowButtons(true), [])
  const onHoverEnd = useCallback(() => setShowButtons(false), [])

  return (
    <div
      className="flex overflow-x-hidden space-x-1"
      onMouseEnter={onHoverStart}
      onMouseLeave={onHoverEnd}
    >
      <span className={className} ref={spanRef}>
        {value}
      </span>
      {unfoldButton}
    </div>
  )
}

interface SectionProps {
  title: string
  entries: [string, string][]
}

const Section = ({ title, entries }: SectionProps) => {
  const rows = useMemo(
    () =>
      entries.map(([name, value]) => (
        <React.Fragment key={`{title.toLowerCase()}-${name}`}>
          <span className="text-right font-medium text-gray-500 select-none">
            {name}
          </span>
          <SectionValue value={value} />
        </React.Fragment>
      )),
    [entries]
  )

  return (
    <>
      <span className="text-lg font-semibold col-span-2 mt-4 select-none">
        {title}
      </span>
      {rows}
    </>
  )
}

interface Props {
  requestId: string
}

const RequestView = ({ requestId }: Props) => {
  const request = useRequest(requestId)

  const url = useMemo(() => new URL(request.url), [request])

  const querySection = useMemo(() => {
    const searchParams = [...url.searchParams.entries()]
    if (searchParams.length === 0) {
      return null
    }

    return <Section title="Query" entries={searchParams}></Section>
  }, [request])

  const headerSection = useMemo(() => {
    const headers = [...Object.entries(request.headers)].filter(
      ([headerName]) => headerName.toLowerCase() !== 'cookie'
    )
    if (headers.length === 0) {
      return null
    }

    return <Section title="Headers" entries={headers}></Section>
  }, [request])

  const cookieSection = useMemo(() => {
    const cookieHeader = [...Object.entries(request.headers)].find(
      ([headerName]) => headerName.toLowerCase() === 'cookie'
    )
    if (cookieHeader === undefined) {
      return null
    }

    const cookies = cookie.parse(cookieHeader[1])

    return (
      <Section title="Cookies" entries={[...Object.entries(cookies)]}></Section>
    )
  }, [request])

  return (
    <div className="max-w-5xl mx-24 mt-8 px-8 pt-2">
      <span className="text-3xl font-bold">{url.host}</span>
      <div className="my-8 grid grid-cols-[10rem_1fr] gap-x-4 gap-y-1">
        {/* URL */}
        <span className="text-right font-medium text-gray-500 select-none">
          URL
        </span>
        <SectionValue value={`${url.origin}${url.pathname}`} />
        {/* Method */}
        <span className="text-right font-medium text-gray-500 select-none">
          Method
        </span>
        <SectionValue value={request.method} />
        {querySection}
        {headerSection}
        {cookieSection}
      </div>
    </div>
  )
}

export default RequestView
