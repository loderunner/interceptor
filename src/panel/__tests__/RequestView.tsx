import { cleanup, fireEvent, render } from '@testing-library/react'
import * as React from 'react'
import '@testing-library/jest-dom'

import RequestView from '../RequestView'

jest.mock('@/interceptor/hooks')

describe('[RequestView]', () => {
  afterEach(() => {
    cleanup()
  })

  it('should match initial snapshot', () => {
    const { container } = render(
      <RequestView requestId={globalMocks.request.id} />
    )
    expect(container).toMatchSnapshot()
  })

  it('should match headers snapshot', () => {
    const { container, getByRole } = render(
      <RequestView requestId={globalMocks.request.id} />
    )

    const headersButton = getByRole('button', { name: 'Headers' })

    fireEvent.click(headersButton)

    expect(container).toMatchSnapshot()
  })

  it('should match body snapshot', () => {
    const { container, getByRole } = render(
      <RequestView requestId={globalMocks.request.id} />
    )

    const bodyButton = getByRole('button', { name: 'Body' })

    fireEvent.click(bodyButton)

    expect(container).toMatchSnapshot()
  })
})
