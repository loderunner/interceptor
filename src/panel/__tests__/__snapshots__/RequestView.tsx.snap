// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[RequestView] should match body snapshot 1`] = `
<div>
  <div
    class="max-w-5xl mx-24 mt-8 px-8 pt-2"
  >
    <div
      class="flex space-x-4 mb-4"
    >
      <button
        class="text-base rounded py-2 px-4 font-medium border hover:bg-gray-300 active:bg-gray-400"
      >
        Headers
      </button>
      <button
        class="text-base rounded py-2 px-4 font-medium border bg-blue-50"
        disabled=""
      >
        Body
      </button>
    </div>
    <div
      class="space-y-1"
    >
      <div
        class="flex items-baseline space-x-1"
        style="margin-left: 0rem;"
      >
        <pre
          class="text-amber-700"
        >
          hello
          :
        </pre>
        <pre
          class="text-sky-400"
        >
          "
          world
          "
        </pre>
      </div>
      <div
        class="flex items-baseline space-x-1"
        style="margin-left: 0rem;"
      >
        <pre
          class="text-amber-700"
        >
          obj
          :
        </pre>
        <pre
          class="flex items-center space-x-0.5"
        >
          {
          <button
            class="rounded-sm bg-slate-400 hover:bg-slate-500 active:bg-slate-600"
          >
            <svg />
          </button>
          }
        </pre>
      </div>
      <div
        class="space-y-1 hidden"
      >
        <div
          class="flex items-baseline space-x-1"
          style="margin-left: 1rem;"
        >
          <pre
            class="text-amber-700"
          >
            barbes
            :
          </pre>
          <pre
            class="text-sky-400"
          >
            "
            rochechouart
            "
          </pre>
        </div>
        <div
          class="flex items-baseline space-x-1"
          style="margin-left: 1rem;"
        >
          <pre
            class="text-amber-700"
          >
            version
            :
          </pre>
          <pre
            class="text-rose-500"
          >
            1
          </pre>
        </div>
        <div
          class="flex items-baseline space-x-1"
          style="margin-left: 1rem;"
        >
          <pre
            class="text-amber-700"
          >
            toto
            :
          </pre>
          <pre
            class="text-purple-500"
          >
            null
          </pre>
        </div>
        <div
          class="flex items-baseline space-x-1"
          style="margin-left: 1rem;"
        >
          <pre
            class="text-amber-700"
          >
            arr
            :
          </pre>
          <pre
            class="flex items-center space-x-0.5"
          >
            [
            <button
              class="rounded-sm bg-slate-400 hover:bg-slate-500 active:bg-slate-600"
            >
              <svg />
            </button>
            ]
          </pre>
        </div>
        <div
          class="space-y-1 hidden"
        >
          <div
            class="flex items-baseline space-x-1"
            style="margin-left: 2rem;"
          >
            <pre
              class="text-amber-700"
            >
              0
              :
            </pre>
            <pre
              class="text-rose-500"
            >
              0.6008246444488461
            </pre>
          </div>
          <div
            class="flex items-baseline space-x-1"
            style="margin-left: 2rem;"
          >
            <pre
              class="text-amber-700"
            >
              1
              :
            </pre>
            <pre
              class="text-rose-500"
            >
              0.39624796882166535
            </pre>
          </div>
          <div
            class="flex items-baseline space-x-1"
            style="margin-left: 2rem;"
          >
            <pre
              class="text-amber-700"
            >
              2
              :
            </pre>
            <pre
              class="text-rose-500"
            >
              0.17061112405361434
            </pre>
          </div>
          <div
            class="flex items-baseline space-x-1"
            style="margin-left: 2rem;"
          >
            <pre
              class="text-amber-700"
            >
              3
              :
            </pre>
            <pre
              class="text-rose-500"
            >
              0.9072284282257375
            </pre>
          </div>
          <div
            class="flex items-baseline space-x-1"
            style="margin-left: 2rem;"
          >
            <pre
              class="text-amber-700"
            >
              4
              :
            </pre>
            <pre
              class="text-rose-500"
            >
              0.03493878762841063
            </pre>
          </div>
        </div>
        <div
          class="flex items-baseline space-x-1"
          style="margin-left: 1rem;"
        >
          <pre
            class="text-amber-700"
          >
            nested
            :
          </pre>
          <pre
            class="flex items-center space-x-0.5"
          >
            {
            <button
              class="rounded-sm bg-slate-400 hover:bg-slate-500 active:bg-slate-600"
            >
              <svg />
            </button>
            }
          </pre>
        </div>
        <div
          class="space-y-1 hidden"
        >
          <div
            class="flex items-baseline space-x-1"
            style="margin-left: 2rem;"
          >
            <pre
              class="text-amber-700"
            >
              thx
              :
            </pre>
            <pre
              class="text-rose-500"
            >
              1138
            </pre>
          </div>
          <div
            class="flex items-baseline space-x-1"
            style="margin-left: 2rem;"
          >
            <pre
              class="text-amber-700"
            >
              star
              :
            </pre>
            <pre
              class="text-sky-400"
            >
              "
              wars
              "
            </pre>
          </div>
        </div>
      </div>
      <div
        class="flex items-baseline space-x-1"
        style="margin-left: 0rem;"
      >
        <pre
          class="text-amber-700"
        >
          done
          :
        </pre>
        <pre
          class="text-green-500"
        >
          true
        </pre>
      </div>
    </div>
  </div>
</div>
`;

exports[`[RequestView] should match headers snapshot 1`] = `
<div>
  <div
    class="max-w-5xl mx-24 mt-8 px-8 pt-2"
  >
    <div
      class="flex space-x-4 mb-4"
    >
      <button
        class="text-base rounded py-2 px-4 font-medium border bg-blue-50"
        disabled=""
      >
        Headers
      </button>
      <button
        class="text-base rounded py-2 px-4 font-medium border hover:bg-gray-300 active:bg-gray-400"
      >
        Body
      </button>
    </div>
    <span
      class="text-3xl font-bold"
    >
      www.example.com
    </span>
    <div
      class="my-8 grid grid-cols-[10rem_1fr] gap-x-4 gap-y-1"
    >
      <span
        class="text-right font-medium text-gray-500 select-none"
      >
        URL
      </span>
      <div
        class="flex overflow-x-hidden space-x-1"
      >
        <span
          class="text-ellipsis overflow-x-hidden whitespace-nowrap"
        >
          https://www.example.com/complete/search
        </span>
      </div>
      <span
        class="text-right font-medium text-gray-500 select-none"
      >
        Method
      </span>
      <div
        class="flex overflow-x-hidden space-x-1"
      >
        <span
          class="text-ellipsis overflow-x-hidden whitespace-nowrap"
        >
          POST
        </span>
      </div>
      <span
        class="text-lg font-semibold col-span-2 mt-4 select-none"
      >
        Query
      </span>
      <span
        class="text-right font-medium text-gray-500 select-none"
      >
        q
      </span>
      <div
        class="flex overflow-x-hidden space-x-1"
      >
        <span
          class="text-ellipsis overflow-x-hidden whitespace-nowrap"
        >
          toto
        </span>
      </div>
      <span
        class="text-right font-medium text-gray-500 select-none"
      >
        client
      </span>
      <div
        class="flex overflow-x-hidden space-x-1"
      >
        <span
          class="text-ellipsis overflow-x-hidden whitespace-nowrap"
        >
          gws-wiz
        </span>
      </div>
      <span
        class="text-lg font-semibold col-span-2 mt-4 select-none"
      >
        Headers
      </span>
      <span
        class="text-right font-medium text-gray-500 select-none"
      >
        Content-Type
      </span>
      <div
        class="flex overflow-x-hidden space-x-1"
      >
        <span
          class="text-ellipsis overflow-x-hidden whitespace-nowrap"
        >
          application/json
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`[RequestView] should match initial snapshot 1`] = `
<div>
  <div
    class="max-w-5xl mx-24 mt-8 px-8 pt-2"
  >
    <div
      class="flex space-x-4 mb-4"
    >
      <button
        class="text-base rounded py-2 px-4 font-medium border hover:bg-gray-300 active:bg-gray-400"
      >
        Headers
      </button>
      <button
        class="text-base rounded py-2 px-4 font-medium border bg-blue-50"
        disabled=""
      >
        Body
      </button>
    </div>
    <div
      class="space-y-1"
    >
      <div
        class="flex items-baseline space-x-1"
        style="margin-left: 0rem;"
      >
        <pre
          class="text-amber-700"
        >
          hello
          :
        </pre>
        <pre
          class="text-sky-400"
        >
          "
          world
          "
        </pre>
      </div>
      <div
        class="flex items-baseline space-x-1"
        style="margin-left: 0rem;"
      >
        <pre
          class="text-amber-700"
        >
          obj
          :
        </pre>
        <pre
          class="flex items-center space-x-0.5"
        >
          {
          <button
            class="rounded-sm bg-slate-400 hover:bg-slate-500 active:bg-slate-600"
          >
            <svg />
          </button>
          }
        </pre>
      </div>
      <div
        class="space-y-1 hidden"
      >
        <div
          class="flex items-baseline space-x-1"
          style="margin-left: 1rem;"
        >
          <pre
            class="text-amber-700"
          >
            barbes
            :
          </pre>
          <pre
            class="text-sky-400"
          >
            "
            rochechouart
            "
          </pre>
        </div>
        <div
          class="flex items-baseline space-x-1"
          style="margin-left: 1rem;"
        >
          <pre
            class="text-amber-700"
          >
            version
            :
          </pre>
          <pre
            class="text-rose-500"
          >
            1
          </pre>
        </div>
        <div
          class="flex items-baseline space-x-1"
          style="margin-left: 1rem;"
        >
          <pre
            class="text-amber-700"
          >
            toto
            :
          </pre>
          <pre
            class="text-purple-500"
          >
            null
          </pre>
        </div>
        <div
          class="flex items-baseline space-x-1"
          style="margin-left: 1rem;"
        >
          <pre
            class="text-amber-700"
          >
            arr
            :
          </pre>
          <pre
            class="flex items-center space-x-0.5"
          >
            [
            <button
              class="rounded-sm bg-slate-400 hover:bg-slate-500 active:bg-slate-600"
            >
              <svg />
            </button>
            ]
          </pre>
        </div>
        <div
          class="space-y-1 hidden"
        >
          <div
            class="flex items-baseline space-x-1"
            style="margin-left: 2rem;"
          >
            <pre
              class="text-amber-700"
            >
              0
              :
            </pre>
            <pre
              class="text-rose-500"
            >
              0.6008246444488461
            </pre>
          </div>
          <div
            class="flex items-baseline space-x-1"
            style="margin-left: 2rem;"
          >
            <pre
              class="text-amber-700"
            >
              1
              :
            </pre>
            <pre
              class="text-rose-500"
            >
              0.39624796882166535
            </pre>
          </div>
          <div
            class="flex items-baseline space-x-1"
            style="margin-left: 2rem;"
          >
            <pre
              class="text-amber-700"
            >
              2
              :
            </pre>
            <pre
              class="text-rose-500"
            >
              0.17061112405361434
            </pre>
          </div>
          <div
            class="flex items-baseline space-x-1"
            style="margin-left: 2rem;"
          >
            <pre
              class="text-amber-700"
            >
              3
              :
            </pre>
            <pre
              class="text-rose-500"
            >
              0.9072284282257375
            </pre>
          </div>
          <div
            class="flex items-baseline space-x-1"
            style="margin-left: 2rem;"
          >
            <pre
              class="text-amber-700"
            >
              4
              :
            </pre>
            <pre
              class="text-rose-500"
            >
              0.03493878762841063
            </pre>
          </div>
        </div>
        <div
          class="flex items-baseline space-x-1"
          style="margin-left: 1rem;"
        >
          <pre
            class="text-amber-700"
          >
            nested
            :
          </pre>
          <pre
            class="flex items-center space-x-0.5"
          >
            {
            <button
              class="rounded-sm bg-slate-400 hover:bg-slate-500 active:bg-slate-600"
            >
              <svg />
            </button>
            }
          </pre>
        </div>
        <div
          class="space-y-1 hidden"
        >
          <div
            class="flex items-baseline space-x-1"
            style="margin-left: 2rem;"
          >
            <pre
              class="text-amber-700"
            >
              thx
              :
            </pre>
            <pre
              class="text-rose-500"
            >
              1138
            </pre>
          </div>
          <div
            class="flex items-baseline space-x-1"
            style="margin-left: 2rem;"
          >
            <pre
              class="text-amber-700"
            >
              star
              :
            </pre>
            <pre
              class="text-sky-400"
            >
              "
              wars
              "
            </pre>
          </div>
        </div>
      </div>
      <div
        class="flex items-baseline space-x-1"
        style="margin-left: 0rem;"
      >
        <pre
          class="text-amber-700"
        >
          done
          :
        </pre>
        <pre
          class="text-green-500"
        >
          true
        </pre>
      </div>
    </div>
  </div>
</div>
`;
